pkg/restore/restore_test.go

```diff
@@ -2113,6 +2112,130 @@ func TestRestorePersistentVolumes(t *testing.T) {
+			podVolumeBackups: []*velerov1api.PodVolumeBackup{
+				builder.ForPodVolumeBackup("velero", "pvb-1").PodName("pod-1").Result(),
+				builder.ForPodVolumeBackup("velero", "pvb-2").PodName("pod-2").Result(),
+			},
+			podWithPVBs: []*corev1api.Pod{
+				builder.ForPod("ns-1", "pod-2").
+					Result(),
+			},
+			podWithoutPVBs: []*corev1api.Pod{
+				builder.ForPod("ns-2", "pod-3").
+					Result(),
+			},
+		},
+		{
+			name:         "a pod that exists in given backup but does not contain associated PVBs should not have any restored pods",
+			restore:      defaultRestore().Result(),
+			backup:       defaultBackup().Result(),
+			apiResources: []*test.APIResource{test.Pods()},
+			podVolumeBackups: []*velerov1api.PodVolumeBackup{
+				builder.ForPodVolumeBackup("velero", "pvb-1").PodName("pod-1").Result(),
+				builder.ForPodVolumeBackup("velero", "pvb-2").PodName("pod-2").Result(),
+			},
+			podWithPVBs: []*corev1api.Pod{},
+			podWithoutPVBs: []*corev1api.Pod{
+				builder.ForPod("ns-1", "pod-3").
+					Result(),
+				builder.ForPod("ns-2", "pod-4").
+					Result(),
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			h := newHarness(t)
+			restorer := new(resticmocks.Restorer)
+			defer restorer.AssertExpectations(t)
+			h.restorer.resticRestorerFactory = &fakeResticRestorerFactory{
+				restorer: restorer,
+			}
+
+			// needed only to indicate resource types that can be restored, in this case, pods
+			for _, resource := range tc.apiResources {
+				h.addItems(t, resource)
+			}
+
+			tarball := newTarWriter(t)
+
+			// these backed up pods don't have any PVBs associated with them, so a call to RestorePodVolumes is not expected to be made for them
+			for _, pod := range tc.podWithoutPVBs {
+				tarball.addItems("pods", pod)
+			}
+
+			// these backed up pods have PVBs associated with them, so a call to RestorePodVolumes will be made for each of them
+			for _, pod := range tc.podWithPVBs {
+				tarball.addItems("pods", pod)
+
+				// the restore process adds these labels before restoring, so we must add them here too otherwise they won't match
+				pod.Labels = map[string]string{"velero.io/backup-name": tc.backup.Name, "velero.io/restore-name": tc.restore.Name}
+				expectedArgs := restic.RestoreData{
+					Restore:          tc.restore,
+					Pod:              pod,
+					PodVolumeBackups: tc.podVolumeBackups,
+					SourceNamespace:  "ns-1",
+					BackupLocation:   "",
+				}
+				restorer.
+					On("RestorePodVolumes", expectedArgs).
+					Return(nil)
+			}
+
+			data := Request{
+				Log:              h.log,
+				Restore:          tc.restore,
+				Backup:           tc.backup,
+				PodVolumeBackups: tc.podVolumeBackups,
+				VolumeSnapshots:  nil,
+				BackupReader:     tarball.done(),
+			}
+
+			warnings, errs := h.restorer.Restore(
+				data,
+				nil, // actions
+				nil, // snapshot location lister
+				nil, // volume snapshotter getter
+			)
+
+			assertEmptyResults(t, warnings, errs)
```

Comment:
it might be good to also ensure that *all* of the pods (whether they have PVBs or not) get restored, to ensure we don't have any test bugs. You can use `assertAPIContents` like one of the other tests for that.